---
title: Explore Chromium molecule size
author: Shaun Jackman
output:
  html_document:
    keep_md: true
params:
  raw_tsv:
    label: "Input raw TSV file of read alignments with columns Rname, Start, End, Size, BX, MI, Reads, Mapq_median, AS_median"
    value: "psitchensiscpmt_2.psitchensis.longranger.wgs.bam.bx.molecule.tsv"
    input: text
  filtered_tsv:
    label: "Input filtered TSV file of read alignments with columns Rname, Start, End, Size, BX, MI, Reads, Mapq_median, AS_median"
    value: "psitchensiscpmt_2.psitchensis.longranger.wgs.bam.as-30.bx.molecule.tsv"
    input: text
  output_breakpoints_tsv:
    label: "Output TSV file of breakpoints"
    value: "psitchensiscpmt_2.breakpoints.tsv"
    input: text
---

```{r setup}
library(cowplot)
library(dplyr)
library(knitr)
library(magrittr)
library(purrr)
library(readr)
library(ggplot2)
library(scales)
library(tidyr)

knit_print.data.frame <- function(x, ...) kable(x) %>% paste(collapse = "\n") %>% asis_output
molecules_filename <- params$raw_tsv
molecules_good_filename <- params$filtered_tsv
output_breakpoints_tsv_filename <- params$output_breakpoints_tsv
```

# Read data
```{r read-data}
molecules_orig <- read_tsv(molecules_filename)
molecules_good_orig <- read_tsv(molecules_good_filename)
```

# Filter
```{r filter}
reads_threshold <- 4
as_median_threshold <- -3
size_threshold <- 500
rname_threshold <- 25

molecules <- molecules_orig %>%
	filter(Reads >= reads_threshold) %>%
	mutate(LogReadDensity = log10(Reads / ifelse(Size == 0, NA, Size)))

molecules_good <- molecules_good_orig %>%
	filter(Reads >= reads_threshold, AS_median >= as_median_threshold, Size >= size_threshold, as.integer(Rname) <= rname_threshold) %>%
	mutate(LogReadDensity = log10(Reads / ifelse(Size == 0, NA, Size)))
```

# Histogram of molecule size
```{r histogram-size}
ggplot(molecules) + aes(x = Size) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = Size, weight = Reads) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Number of reads", labels = comma)

ggplot(molecules) + aes(x = Size, weight = Size) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Total DNA mass (Mbp)", labels = unit_format(unit = "Mbp", scale = 1e-6))
```

# Histogram of reads per molecule
```{r histogram-reads}
ggplot(molecules) + aes(x = Reads) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(limits = c(0, 50)) +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = Reads, weight = Reads) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(limits = c(0, 50)) +
	scale_y_continuous(name = "Number of reads", labels = comma)
```

# Histogram of alignment score (AS)
```{r histogram-as}
ggplot(molecules) + aes(x = AS_median) +
	geom_histogram(binwidth = 2, boundary = 0) +
	scale_x_continuous(name = "Median alignment score") +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = AS_median, weight = Reads) +
	geom_histogram(binwidth = 2, boundary = 0) +
	scale_x_continuous(name = "Median alignment score") +
	scale_y_continuous(name = "Number of reads", labels = comma)
```

# Histogram of mapping quality (MAPQ)
```{r histogram-mapq}
ggplot(molecules) + aes(x = Mapq_median) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = Mapq_median, weight = Reads) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_y_continuous(name = "Number of reads", labels = comma)
```

# Histogram of molecules per barcode
```{r histogram-molecules-per-barcode}
molecules_per_barcode <- molecules %>% count(BX) %>% rename(Molecules = n)

ggplot(molecules_per_barcode) + aes(x = Molecules) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(name = "Number of molecules per barcode") +
	scale_y_continuous(name = "Number of barcodes", labels = comma)
```

# Read density per molecule
```{r read-density}
ggplot(molecules) +
	aes(x = LogReadDensity) +
	geom_histogram(boundary = 0, binwidth = 0.1) +
	geom_vline(data = tibble(LogReadDensity = boxplot.stats(molecules$LogReadDensity)$stats), aes(xintercept = LogReadDensity),
		colour = "blue", linetype = "dashed", alpha = 0.5)

ggplot(molecules_good) +
	aes(x = LogReadDensity) +
	geom_histogram(boundary = 0, binwidth = 0.1) +
	geom_vline(data = tibble(LogReadDensity = boxplot.stats(molecules_good$LogReadDensity)$stats), aes(xintercept = LogReadDensity),
		colour = "blue", linetype = "dashed", alpha = 0.5) +
	labs(caption = "Filtered by Size and AS")
```

# Heatmap of molecule size and number of reads per molecule
```{r heatmap-reads-size}
ggplot(molecules) + aes(x = Size, y = Reads) +
	geom_hex() +
	geom_smooth() +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Number of reads per molecule", labels = comma) +
	scale_fill_continuous("Number of molecules", trans = log10_trans(), labels = comma)
```

# Heatmap of molecule size and read density per molecule
```{r heatmap-size-density}
ggplot(molecules) + aes(x = Size, y = LogReadDensity) +
	geom_hex() +
	geom_smooth() +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Log10 read density", labels = comma) +
	scale_fill_continuous("Number of molecules", trans = log10_trans(), labels = comma)
```

# Heatmap of number of reads and read density per molecule
```{r heatmap-reads-density}
ggplot(molecules) + aes(x = Reads, y = LogReadDensity) +
	geom_hex() +
	geom_smooth() +
	scale_x_log10(name = "Number of reads per molecule", labels = comma) +
	scale_y_continuous(name = "Log10 read density", labels = comma) +
	scale_fill_continuous("Number of molecules", trans = log10_trans(), labels = comma)
```

# Identify misassemblies
```{r identify-misassemblies}
depth_good_threshold <- 75
pos_threshold <- 2500
starts_as_threshold <- -30
starts_size_threshold <- 2000
starts_threshold <- 4

molecules_good_pos <- rbind(
		transmute(molecules_good, Rname, Pos = Start),
		transmute(molecules_good, Rname, Pos = End)) %>%
	distinct()

molecules_depth_good <- molecules_good_pos %>% rename(Pos_Rname = Rname) %>% rowwise() %>%
	do(Rname = .$Pos_Rname, Pos = .$Pos, Sizes = filter(molecules_good, Rname == .$Pos_Rname & Start <= .$Pos & End > .$Pos)$Size) %>%
	mutate(Depth = length(Sizes), Median_size = as.numeric(median(Sizes))) %>%
	replace_na(list(Median_size = 0)) %>%
	unnest(Rname, Pos) %>%
	select(Rname, Pos, Depth, Median_size, everything()) %>%
	arrange(Rname, Pos)

molecules_starts <- molecules %>%
	filter(as.integer(Rname) <= rname_threshold, AS_median >= starts_as_threshold & Size >= starts_size_threshold) %>%
	group_by(Rname, Start) %>%
	summarize(Starts = n(), Starts_median_size = median(Size), Starts_AS_median = median(AS_median), Starts_sum_reads = sum(Reads)) %>%
	ungroup() %>%
	rename(Pos = Start) %>%
	full_join(molecules_depth_good, by = c("Rname", "Pos")) %>%
	arrange(Rname, Pos) %>%
	fill(Depth, Median_size) %>%
	select(Rname, Pos, Depth, Median_size, everything())

misassemblies <- molecules_starts %>%
	filter(Pos >= pos_threshold & Starts >= starts_threshold & Depth < depth_good_threshold) %>%
	select(-Sizes)

misassemblies %>% arrange(as.integer(Rname))
```

# Plot misassemblies, depth of molecule coverage and molecule starts
```{r misassemblies-depth, fig.width=7, fig.height=7}
ggplot(molecules_starts) +
	facet_wrap(~ Rname, scales = "free_x") +
	aes(x = Pos) +
	geom_point(aes(x = Pos, y = Depth), size = 0.5) +
	geom_point(aes(y = Starts), size = 0.5) +
	geom_hline(yintercept = depth_good_threshold, colour = "orange", linetype = "dashed") +
	geom_vline(aes(xintercept = Pos), misassemblies, colour = "blue", linetype = "dashed") +
	scale_x_continuous(labels = unit_format(unit = "", scale = 1e-3))
```

# Plot misassemblies and scatter plot of start and end coordinates
```{r misassemblies-start-end, fig.width=7, fig.height=7}
molecules_good %>% filter(as.integer(Rname) <= rname_threshold) %>% ggplot() +
	aes(x = Start, y = End, alpha = Reads) +
	geom_point(size = 1) +
	geom_vline(aes(xintercept = Pos), misassemblies, colour = "blue", linetype = "dashed") +
	geom_hline(aes(yintercept = Pos), misassemblies, colour = "blue", linetype = "dashed") +
	facet_wrap(~ Rname, scales = "free") +
	scale_x_continuous(labels = unit_format(unit = "", scale = 1e-3)) +
	scale_y_continuous(labels = unit_format(unit = "", scale = 1e-3))

ggplot(molecules_good %>% filter(as.integer(Rname) <= rname_threshold)) +
	aes(x = Start, y = End) +
	geom_bin2d(binwidth = 2000) +
	geom_vline(aes(xintercept = Pos), misassemblies, colour = "blue", linetype = "dashed") +
	geom_hline(aes(yintercept = Pos), misassemblies, colour = "blue", linetype = "dashed") +
	facet_wrap(~ Rname, scales = "free") +
	scale_x_continuous(labels = unit_format(unit = "", scale = 1e-3)) +
	scale_y_continuous(labels = unit_format(unit = "", scale = 1e-3))
```

# Plot misassemblies and molecule start and end coordinates of a scaffold
```{r ggplot-misassemblies}
ggplot_misassemblies <- function(scaffold_name) {
	max_x <- max(
		filter(molecules, Rname == scaffold_name)$End,
		filter(molecules_starts, Rname == scaffold_name)$Pos)

	scatter_plot <- molecules_good %>% filter(Rname == scaffold_name) %>% ggplot() +
		aes(x = Start, y = End) +
		geom_point(size = 1) +
		geom_vline(aes(xintercept = Pos), misassemblies %>% filter(Rname == scaffold_name), colour = "blue", linetype = "dashed") +
		geom_hline(aes(yintercept = Pos), misassemblies %>% filter(Rname == scaffold_name), colour = "blue", linetype = "dashed") +
		scale_x_continuous(labels = unit_format(unit = "kbp", scale = 1e-3), limits = c(0, max_x)) +
		scale_y_continuous(labels = unit_format(unit = "kbp", scale = 1e-3), limits = c(0, max_x))

	plot <- molecules_starts %>% filter(Rname == scaffold_name) %>% ggplot() +
		geom_vline(aes(xintercept = Pos), misassemblies %>% filter(Rname == scaffold_name), colour = "blue", linetype = "dashed") +
		scale_x_continuous(labels = unit_format(unit = "kbp", scale = 1e-3), limits = c(0, max_x))

	depth_plot <- plot + geom_point(aes(x = Pos, y = Depth), size = 1) +
		geom_hline(yintercept = depth_good_threshold, colour = "orange", linetype = "dashed")

	starts_plot <- plot + geom_point(aes(x = Pos, y = Starts), size = 1) +
		geom_hline(yintercept = starts_threshold, colour = "orange", linetype = "dashed")

	plot_grid(scatter_plot, depth_plot, starts_plot, ncol = 1, align = "v", rel_heights = c(3, 1, 1), labels = scaffold_name)
}
```

# Plot misassemblies and molecule start and end coordinates of a scaffold
```{r plot-misassemblies}
misassemblies %>%
	arrange(as.integer(Rname)) %$%
	unique(Rname) %>%
	map(ggplot_misassemblies)
```

# Count misasemblies per scaffold
```{r misassemblies-per-scaffold}
misassemblies %>% count(Rname) %>% arrange(as.integer(Rname))
```

# Store molecule depth and misassemblies in a TSV file
```{r write-tsv}
molecules_starts %>% select(-Sizes) %>% arrange(as.integer(Rname), Pos) %>% write_tsv("molecule-depth.tsv")
misassemblies %>% arrange(as.integer(Rname), Pos) %>% write_tsv(output_breakpoints_tsv_filename)
```
