---
title: Explore Chromium molecule size
author: Shaun Jackman
output:
  html_document:
    keep_md: true
params:
  input_tsv:
    label: "Input TSV file of read alignments with columns Rname, Start, End, Size, BX, MI, Reads, Mapq_median, AS_median"
    value: "psitchensiscpmt_2.psitchensis.longranger.wgs.bam.bx.molecule.tsv"
    input: text
---

```{r setup}
library(dplyr)
library(knitr)
library(magrittr)
library(purrr)
library(readr)
library(ggplot2)
library(scales)
library(tidyr)

knit_print.data.frame <- function(x, ...) kable(x) %>% paste(collapse = "\n") %>% asis_output
molecules_filename <- params$input_tsv
```

# Read data
```{r read-data}
molecules_orig <- read_tsv(molecules_filename)
```

# Filter
```{r filter}
count_threshold <- 4
molecules <- molecules_orig %>% filter(Reads >= count_threshold)
```

# Histogram of molecule size
```{r histogram-size}
ggplot(molecules) + aes(x = Size) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = Size, weight = Reads) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Number of reads", labels = comma)

ggplot(molecules) + aes(x = Size, weight = Size) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Total DNA mass (Mbp)", labels = unit_format(unit = "Mbp", scale = 1e-6))

ggplot(molecules) + aes(x = Size, weight = Reads) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(limits = c(1500, 50000), name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_continuous(name = "Number of reads", labels = comma)
```

# Histogram of reads per molecule
```{r histogram-reads}
ggplot(molecules) + aes(x = Reads) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(limits = c(0, 50)) +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = Reads, weight = Reads) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(limits = c(0, 50)) +
	scale_y_continuous(name = "Number of reads", labels = comma)
```

# Histogram of alignment score (AS)
```{r histogram-as}
ggplot(molecules) + aes(x = AS_median) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(limits = c(-150, 0)) +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = AS_median, weight = Reads) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(limits = c(-150, 0)) +
	scale_y_continuous(name = "Number of reads", labels = comma)
```

# Histogram of mapping quality (MAPQ)
```{r histogram-mapq}
ggplot(molecules) + aes(x = Mapq_median) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_y_continuous(name = "Number of molecules", labels = comma)

ggplot(molecules) + aes(x = Mapq_median, weight = Reads) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_y_continuous(name = "Number of reads", labels = comma)
```

# Histogram of molecules per barcode
```{r histogram-molecules-per-barcode}
molecules_per_barcode <- molecules %>% count(BX) %>% rename(Molecules = n)

ggplot(molecules_per_barcode) + aes(x = Molecules) +
	geom_histogram(binwidth = 1, boundary = 0) +
	scale_x_continuous(name = "Number of molecules per barcode") +
	scale_y_continuous(name = "Number of barcodes", labels = comma)
```

# Histogram of DNA per barcode
```{r histogram-dna-per-barcode}
dna_per_barcode <- molecules %>% group_by(BX) %>% summarize(DNA = sum(Size))

ggplot(dna_per_barcode) + aes(x = DNA) +
	geom_histogram(binwidth = 500, boundary = 0) +
	scale_x_continuous(name = "DNA per barcode") +
	scale_y_continuous(name = "Number of barcodes", labels = comma)
```

# Heatmap of molecule size and number of reads per molecule
```{r heatmap-reads-size}
ggplot(molecules) + aes(x = Size, y = Reads) + geom_hex() +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_log10(name = "Number of reads per molecule", labels = comma) +
	scale_fill_continuous("Number of molecules", trans = log10_trans(), labels = comma)

ggplot(molecules) + aes(x = Size, y = Reads, weight = Reads) + geom_hex() +
	scale_x_continuous(name = "Molecule size", labels = unit_format(unit = "kbp", scale = 1e-3)) +
	scale_y_log10(name = "Number of reads per molecule", labels = comma) +
	scale_fill_continuous("Number of reads", trans = log10_trans(), labels = comma)
```

# Scaffold 4

# Scatter plot of start and end coordinates
```{r scatter-start-end}
as_median_threshold <- -4
size_threshold <- 500

scaffold4_all <- molecules %>% filter(Rname == "4")
scaffold4_good <- molecules %>% filter(Rname == "4" & AS_median > as_median_threshold & Size >= size_threshold)
scaffold4 <- scaffold4_all
scaffold4 %>% filter(Start < 108910 & End > 109284)

ggplot(scaffold4) + aes(x = Start, y = End, alpha = Reads) + geom_point()
ggplot(scaffold4) + aes(x = Start, y = Size) + geom_point()
```

# Histogram of start coordinates
```{r histogram-start}
starts_size_threshold <- 2000
ggplot(scaffold4 %>% filter(Size >= starts_size_threshold)) + aes(x = Start) + geom_histogram(binwidth = 1) + xlim(108500, 109500)
scaffold4_starts <- scaffold4 %>%
	filter(Size >= starts_size_threshold) %>%
	group_by(Rname, Start) %>%
	summarize(Starts = n(), AS_median = median(AS_median), Reads = sum(Reads)) %>%
	ungroup() %>%
	rename(Pos = Start) %>%
	arrange(desc(Starts)) %>% head
scaffold4_starts %>% arrange(desc(Starts)) %>% head
```

# Depth of molecule coverage
```{r depth-of-coverage}
scaffold4_pos <- unique(c(scaffold4$Start, scaffold4$End))
depth <- scaffold4_pos %>% map(~ filter(scaffold4, Start <= .x & End > .x) %>% nrow()) %>% unlist
scaffold4_depth <- data_frame(Pos = scaffold4_pos, Depth = depth)
ggplot(scaffold4_depth) + aes(x = Depth) + geom_histogram(binwidth = 1)
ggplot(scaffold4_depth) + aes(x = Pos, y = Depth, colour = Depth < 100) + geom_point()
scaffold4_depth %>% filter(Depth < 100 & Pos > 2000 & Pos < 117000) %>% summarize(min(Pos), max(Pos))
```

# Median molecule size spanning each position
```{r median-size-all}
scaffold4_pos <- unique(c(scaffold4$Start, scaffold4$End))
median_size <- scaffold4_pos %>% map(~ filter(scaffold4, Start <= .x & End > .x) %$% median(Size)) %>% unlist
scaffold4_median_size <- data_frame(Pos = scaffold4_pos, Median_size = median_size) %>% replace_na(list(Median_size = 0))
ggplot(scaffold4_median_size) + aes(x = Median_size) + geom_histogram(binwidth = 1000, boundary = 0)
ggplot(scaffold4_median_size) + aes(x = Pos, y = Median_size, colour = Median_size < 10000) + geom_point()
scaffold4_median_size %>% filter(Median_size < 10000 & Pos > 10000 & Pos < 110000) %>% summarize(min(Pos), max(Pos))
```

# Median molecule size spanning each position (filtered by Size and AS)
```{r median-size-good}
median_size_threshold <- 10000
scaffold4_good_pos <- unique(c(scaffold4_good$Start, scaffold4_good$End))
median_size <- scaffold4_good_pos %>% map(~ filter(scaffold4_good, Start <= .x & End > .x) %$% median(Size)) %>% unlist
scaffold4_median_size <- data_frame(Pos = scaffold4_good_pos, Median_size = median_size) %>% replace_na(list(Median_size = 0))
ggplot(scaffold4_median_size) + aes(x = Median_size) + geom_histogram(binwidth = 1000, boundary = 0)
ggplot(scaffold4_median_size) + aes(x = Pos, y = Median_size, colour = Median_size < median_size_threshold) + geom_point()
scaffold4_median_size %>% filter(Median_size < median_size_threshold & Pos > 10000 & Pos < 110000) %>% summarize(min(Pos), max(Pos))
```

# Depth of molecule coverage (filtered by Size and AS)
```{r depth-of-coverage-good}
depth_good_threshold <- 50
depth_good <- scaffold4_good_pos %>% map(~ filter(scaffold4_good, Start <= .x & End > .x) %>% nrow()) %>% unlist
scaffold4_depth_good <- data_frame(Pos = scaffold4_good_pos, Depth = depth_good)
ggplot(scaffold4_depth_good) + aes(x = Depth) + geom_histogram(binwidth = 1)
ggplot(scaffold4_depth_good) + aes(x = Pos, y = Depth, colour = Depth < depth_good_threshold) + geom_point()
```

# Identify misassemblies
```{r misassemblies}
scaffold4_features <- scaffold4_starts %>%
	full_join(scaffold4_depth_good, by = "Pos") %>%
	full_join(scaffold4_median_size, by = "Pos") %>%
	select(Rname, Pos, everything()) %>%
	arrange(Pos) %>%
	fill(Depth, Median_size)

starts_threshold <- 3
scaffold4_misassemblies <-  scaffold4_features %>%
	filter(Starts > starts_threshold & Depth < depth_good_threshold)
scaffold4_misassemblies

ggplot(scaffold4_features) +
	aes(x = Pos) +
	geom_point(aes(y = Median_size / 1000, colour = Median_size < 10000)) +
	geom_point(aes(x = Pos, y = Depth, colour = Depth < depth_good_threshold)) +
	geom_point(aes(y = Starts)) +
	geom_vline(aes(xintercept = Pos), scaffold4_misassemblies)
```

# Store misassemblies in a TSV file
```{r misassemblies_tsv}
write_tsv(scaffold4_misassemblies, "misassemblies.tsv")
```
